`rptest`: transactional control batch testing follow-ups

WillemKauf opened the pull request:
Sequel to https://github.com/redpanda-data/redpanda/pull/28205.

## Backports Required

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

* none

WillemKauf committed:
`rptest`: upgrade to `HEAD` in `LogCompactionTxRemovalUpgradeTest`

WillemKauf committed:
`rptest`: split `TxUpgradeTest` into `Base` class

WillemKauf committed:
`rptest`: add `TxUpgradeCompactionTest`

WillemKauf committed:
`rptest`: `ruff` linting

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--2f8d8763-0730-47ee-a827-39dde393dfaf-->
<!--TEST RESULTS-->
<!--6b4bb050-c3bf-456e-adf7-d571d4b966d5-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/75063">build#75063</a></summary>


| test_class                 | test_method                     | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                            |
|:---------------------------|:--------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ConsumerGroupBalancingTest | test_coordinator_nodes_balance  | null             | integration | https://buildkite.com/redpanda/redpanda/builds/75063#019a2739-ab13-421a-96b1-fe2b579b4881 | FLAKY         | 20/21    | upstream reliability is '99.77678571428571'. current run reliability is '95.23809523809523'. drift is 4.53869 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ConsumerGroupBalancingTest&test_method=test_coordinator_nodes_balance |
| DataMigrationsApiTest      | test_higher_level_migration_api | null             | integration | https://buildkite.com/redpanda/redpanda/builds/75063#019a2740-b2ae-49fb-9933-095f8ecf64f4 | FLAKY         | 19/21    | upstream reliability is '99.43074003795066'. current run reliability is '90.47619047619048'. drift is 8.95455 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DataMigrationsApiTest&test_method=test_higher_level_migration_api     |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/75179">build#75179</a></summary>


| test_class                       | test_method                                | test_arguments                                                     | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                               |
|:---------------------------------|:-------------------------------------------|:-------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DeleteRecordsTest                | test_delete_records_concurrent_truncations | {"cloud_storage_enabled": true, "truncate_point": "random_offset"} | integration | https://buildkite.com/redpanda/redpanda/builds/75179#019a2e6e-06eb-4d74-bde4-a0884161410e | FLAKY         | 17/21    | upstream reliability is '100.0'. current run reliability is '80.95238095238095'. drift is 19.04762 and the allowed drift is set to 50. The test should PASS            | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DeleteRecordsTest&test_method=test_delete_records_concurrent_truncations |
| WriteCachingFailureInjectionTest | test_unavoidable_data_loss                 | null                                                               | integration | https://buildkite.com/redpanda/redpanda/builds/75179#019a2e6c-7357-44bc-8e58-9283537c68b5 | FLAKY         | 19/21    | upstream reliability is '93.45661450924608'. current run reliability is '90.47619047619048'. drift is 2.98042 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionTest&test_method=test_unavoidable_data_loss  |
</details>
<!--/body-->

bharathv commented on tests/rptest/transactions/tx_upgrade_test.py:
```diff
@@ -124,6 +142,78 @@ def upgrade_does_not_change_tx_coordinator_assignment_test(self):
+            for node in other_nodes:
+                self.upgrade_one_node(node, new_version)
+            unique_versions = wait_for_num_versions(self.redpanda, 1)
+            assert prev_version_str not in unique_versions, unique_versions
+            assert self._get_tx_id_mapping() == initial_mapping, (
+                "Mapping changed after full upgrade"
+            )
+            prev_version_str = ver_string(new_version)
```
do we need any more validation to ensure all control batches are removed?

WillemKauf replied:
Added validation here

WillemKauf committed:
`kgo-verifier`: add support for custom `transaction.timeout.ms`

WillemKauf committed:
`rptest`: add `wait_for_acks` setting to `KgoVerifierProducer`

In some cases, when we expect not all of our messages to be properly
acked (e.g. if a partition movement occurs during a transaction), we
may want to wait for the number of messages sent to be what we expect
instead of the number of messages acked.

WillemKauf committed:
`rptest`: deflake `LogCompactionTxRemovalTest`

1. An open transaction could hold back LSO in `rm_stm`, and therefore
hold back compaction. Set `transaction.timeout.ms` to a low value and
sleep before waiting on compaction to settle to prevent this from happening.
2. A partition movement during a transaction could result in the transaction
failing, and the producer never reaching the right number of messages
acked. Use `wait_for_acks=False` with `KgoVerifierProducer` to avoid this
situation (and allow some data to be dropped).
3. Set `tolerate_data_loss=True`. This seems to be necessary, unfortunately.

WillemKauf committed:
`tools`: make `offline_log_viewer` output valid JSON

Printing out a topic with more than one partition will result in a raised
`JSON` error within `offline_log_viewer.py` in ducktape, since previously
partitions were printed as the following:

```
[
  {partition_1}
]
[
  {partition_2}
]
```

Note the multiple top-level arrays with no `,` delimiter in between.
Now, the output will be:

```
[
  [
    {partition_1}
  ],
  [
    {partition_2}
  ]
]
```

which is valid JSON.

WillemKauf committed:
`cluster`: alter `vassert` condition in `reset_with_new_epoch()`

We remove `tx_fence` batches during compaction. It is a valid operation
to update the epoch when not currently in an open transaction.

Without this change, we may hit this `vassert()` when `tx_fence` batches
have been compacted away, and an epoch bump is requested due to applied
transactional batches in `producer_state::apply_data()`.
